### PR TITLE
feat: cleanup indexer transaction.ts

### PR DESCRIPTION
**Pull request by dznes**
Reorganize and cleanup indexer transaction.ts

Time spent on this PR: 3 hours

Resolves: #1221

# Pull Request type
Reorganize and cleanup indexer transaction.ts from issue #1221 

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [X] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [X] Documentation content changes
- [ ] Testing

#Detailed Description
The following changes have been made to the indexer transaction.ts file:

Comments and Documentation: Added detailed comments and documentation to explain the purpose of each function, its parameters, and its return values.
Error Handling: Improved error handling by adding clear and descriptive error messages.
Consistent Formatting: Ensured consistent formatting with proper indentation and spacing.
Descriptive Variable Names: Used meaningful and descriptive variable names for better readability.
Code Organization: Grouped related functions together logically and organized the code for better structure.
Simplified Logic: Refactored complex logic into smaller, more manageable functions.

# What is the new behavior?
None

# Does this introduce a breaking change?

- [ ] Yes
- [X] No

### CODE REVIEW

File: indexer/src/types/transaction.ts [hunk 1]
```diff
@@ -28,35 +28,68 @@ import {
...
+ * @param transaction - The transaction object to be converted.
+ * @param receipt - The transaction receipt object.
  * @param blockNumber - The block number of the transaction in hex.
  * @param blockHash - The block hash of the transaction in hex.
  * @param isPendingBlock - Whether the block is pending.
  * @returns - The transaction in the Ethereum format, or null if the transaction is invalid.
- *
- * Acknowledgement: Code taken from <https://github.com/ethereumjs/ethereumjs-monorepo>
+ * 
+ *  * Acknowledgement: Code taken from <https://github.com/ethereumjs/ethereumjs-monorepo>
```
[the remark this answers is not in the document]
Comment (dznes) [reply indexer/src/types/transaction.ts:79]: It was like this before, but I can remove it.

**Comment by dznes** [comment]
hello! 
I've made the asked changes. If there is anything more, please fell free to add the comments!

File: indexer/src/types/transaction.ts [hunk 2]
```diff
@@ -28,35 +26,69 @@ import {
...
+  blockHash: PrefixedHexString;
+  isPendingBlock: boolean;
+}
+
+interface ExtendedJsonRpcTx extends JsonRpcTx {
+  yParity?: string,
+  isRunOutOfResources?: boolean 
+}
+
+interface TypedTransactionToEthTxRequest {
```
Comment (tcoratger) [inline indexer/src/types/transaction.ts:49]: This naming seems too long and confusing to me, we should keep names as short as possible

File: indexer/src/types/transaction.ts [hunk 3]
```diff
@@ -28,35 +26,69 @@ import {
...
 
+type HexString = `0x${string}`
+
+interface ToEthTxRequest {
+  transaction: Transaction;
+  receipt: TransactionReceipt;
+  blockNumber: PrefixedHexString;
+  blockHash: PrefixedHexString;
+  isPendingBlock: boolean;
+}
```
Comment (tcoratger) [inline indexer/src/types/transaction.ts:42]: In all the interfaces like this one, can we have a header documentation for the interface as well as for the fields inside in order to document everything properly and for a new user to understand each line of the code

File: indexer/src/types/transaction.ts [hunk 4]
```diff
@@ -67,28 +99,138 @@ export function toEthTx({
...
+function buildTransactionEthFormat({
+  typedTransaction,
+  jsonTx,
+  blockNumber,
+  blockHash,
+  isPendingBlock,
+  chainId,
+  index,
+}: BuildTransactionEthFormat): ExtendedJsonRpcTx | null {
+  if (jsonTx.v === undefined || jsonTx.r === undefined || jsonTx.s === undefined) {
```
Comment (tcoratger) [inline indexer/src/types/transaction.ts:185]:
```suggestion
  if (!jsonTx.v || !jsonTx.r || !jsonTx.s) {
```

File: indexer/src/types/transaction.ts [hunk 5]
```diff
@@ -67,28 +99,138 @@ export function toEthTx({
...
+    transactionIndex: isPendingBlock ? null : padBigint(BigInt(index ?? 0), 8),
+    value: jsonTx.value!,
+    v: jsonTx.v,
+    r: jsonTx.r,
+    s: jsonTx.s,
+    maxFeePerBlobGas: jsonTx.maxFeePerBlobGas,
+    blobVersionedHashes: jsonTx.blobVersionedHashes,
+  };  
+
+  return result;
```
Comment (tcoratger) [inline indexer/src/types/transaction.ts:214]:
```suggestion
  return {
    blockHash: isPendingBlock ? null : blockHash,
    blockNumber,
    from: typedTransaction.getSenderAddress().toString(), // no need to pad as the `Address` type is 40 bytes.
    gas: jsonTx.gasLimit!,
    gasPrice: jsonTx.gasPrice ?? jsonTx.maxFeePerGas!,
    maxFeePerGas: jsonTx.maxFeePerGas,
    maxPriorityFeePerGas: jsonTx.maxPriorityFeePerGas,
    type: intToHex(typedTransaction.type),
    accessList: jsonTx.accessList,
    chainId,
    hash: padBytes(typedTransaction.hash(), 32),
    input: jsonTx.data!,
    nonce: jsonTx.nonce!,
    to: typedTransaction.to?.toString() ?? null,
    transactionIndex: isPendingBlock ? null : padBigint(BigInt(index ?? 0), 8),
    value: jsonTx.value!,
    v: jsonTx.v,
    r: jsonTx.r,
    s: jsonTx.s,
    maxFeePerBlobGas: jsonTx.maxFeePerBlobGas,
    blobVersionedHashes: jsonTx.blobVersionedHashes,
  };  
```

File: indexer/src/types/transaction.ts [hunk 6]
```diff
@@ -28,35 +26,69 @@ import {
...
   TypedTransaction,
   TypedTxData,
+  hexToBytes,
+  bytesToBigInt,
+  JsonTx,
 } from "../deps.ts";
 
+type HexString = `0x${string}`
+
+interface ToEthTxRequest {
```
Comment (tcoratger) [inline indexer/src/types/transaction.ts:36]: Can we create a separate dedicated file where we store all the interfaces and types as they are constants and can be reused for other purposes?

File: indexer/src/types/transaction.ts [hunk 7]
```diff
@@ -67,28 +99,138 @@ export function toEthTx({
...
+ * @param typedTransaction - The typed transaction object.
+ * @param jsonTx - The JSON representation of the transaction.
+ * @returns - The chain ID as a string, or undefined if the chain ID cannot be determined.
+ *
+ * If the transaction is a legacy transaction that supports EIP-155 replay protection,
+ * the chain ID is calculated from the `v` value using the formula:
+ * v = 35 + 2 * chainId + yParity -> chainId = (v - 35) / 2
+ */
+function calculateChainId(typedTransaction: TypedTransaction, jsonTx: JsonTx): string | undefined {
+  if (isLegacyTx(typedTransaction) && typedTransaction.supports(Capability.EIP155ReplayProtection)) {
```
Comment (tcoratger) [inline indexer/src/types/transaction.ts:113]: you can add

```typescript
  const { v, chainId } = jsonTx;
```

at the beginning of the function to have clear reusable components

File: indexer/src/types/transaction.ts [hunk 8]
```diff
@@ -67,28 +99,138 @@ export function toEthTx({
...
+    if (jsonTx.v !== undefined) {
+      return bigIntToHex((BigInt(jsonTx.v) - 35n) / 2n);
+    } else {
+      console.error("jsonTx.v is undefined");
+      return undefined;
+    }
```
Comment (tcoratger) [inline indexer/src/types/transaction.ts:119]:
```suggestion
return jsonTx.v ?  bigIntToHex((BigInt(jsonTx.v) - 35n) / 2n) : (console.error("jsonTx.v is undefined"), undefined);
```

File: indexer/src/types/transaction.ts [hunk 9]
```diff
@@ -67,28 +99,138 @@ export function toEthTx({
...
+}
+
+/**
+ * Adds the yParity field to the transaction result if it is an EIP-1559 or EIP-2930 transaction.
+ *
+ * @param typedTransaction - The typed transaction object.
+ * @param jsonTx - The JSON representation of the transaction.
+ * @param result - The transaction result object in Ethereum format.
+ */
+function addYParityIfNeeded(
```
Comment (tcoratger) [inline indexer/src/types/transaction.ts:131]: this function name seems strange to me

File: indexer/src/types/transaction.ts [hunk 10]
```diff
@@ -67,28 +99,138 @@ export function toEthTx({
...
+}
+
+/**
+ * Adds the isRunOutOfResources flag to the transaction result if the transaction
+ * was reverted due to running out of resources.
+ *
+ * @param receipt - The transaction receipt object.
+ * @param result - The transaction result object in Ethereum format.
+ */
+function addRunOutOfResourcesFlagIfNeeded(
```
Comment (tcoratger) [inline indexer/src/types/transaction.ts:153]: Same, this function name seems strange to me

File: indexer/src/types/transaction.ts [hunk 11]
```diff
@@ -270,20 +356,24 @@ function fromSerializedData(bytes: Uint8Array): TypedTransaction {
...
     // the chain id to be >= 37. This value will be updated during the call to addSignature.
     values[6] = bigIntToBytes(37n);
-    return LegacyTransaction.fromValuesArray(
+
+    const typedTransaction = LegacyTransaction.fromValuesArray(
       values as TxValuesArray[TransactionType.Legacy],
-    );
+    )
+    
+    return typedTransaction;
```
Comment (tcoratger) [inline indexer/src/types/transaction.ts:364]: Why do we need to change that? I liked the previous compact format more

File: indexer/src/types/transaction.ts (hunk 6 above)
Comment (dznes) [reply indexer/src/types/transaction.ts:36]: No problem, should I create a file like "interfaces.ts" or keep all the interfaces in the "dep.ts" file with the other interfaces?

File: indexer/src/types/transaction.ts (hunk 2 above)
Comment (dznes) [reply indexer/src/types/transaction.ts]: changed to "TypedTxToEthTx"

File: indexer/src/types/transaction.ts (hunk 3 above)
Comment (dznes) [reply indexer/src/types/transaction.ts]: added headers documentation and fields description

File: indexer/src/types/transaction.ts (hunk 9 above)
Comment (dznes) [reply indexer/src/types/transaction.ts]: changed to "setYParityFlag", please give me a sugestion if the new one doesnt fit.

File: indexer/src/types/transaction.ts (hunk 10 above)
Comment (dznes) [reply indexer/src/types/transaction.ts]: changed to "flagResourceExhaustion", please give me a sugestion if the new one doesnt fit.

File: indexer/src/types/transaction.ts (hunk 11 above)
Comment (dznes) [reply indexer/src/types/transaction.ts]: changed back to the original code

File: indexer/src/types/transaction.ts (hunk 6 above)
Comment (dznes) [reply indexer/src/types/transaction.ts]: Created a file named "interfaces.ts" in the same folder. If this doesn't match what is needed, please specify the correct location for the file.

File: indexer/src/types/interfaces.ts [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+
+/**
+ * Represents a request to convert a transaction to Ethereum transaction format.
+ */
+export interface ToEthTxRequest {
+  transaction: Transaction; //  A Ethereum transaction.
+  receipt: TransactionReceipt; // The Ethereum receipt corresponding to a reverted out of resources transaction.
+  blockNumber: PrefixedHexString; // The block number in which the transaction was included, as a prefixed hex string.
+  blockHash: PrefixedHexString; // The hash of the block in which the transaction was included, as a prefixed hex string.
+  isPendingBlock: boolean; // Indicates if the transaction is in a pending block.
```
Comment (tcoratger) [inline indexer/src/types/interfaces.ts:25]: I meant the documentation header which is displayed correctly above the field in question for example:

```ts
export interface ToEthTxRequest {
/** An Ethereum transaction. */
 transaction: Transaction;
```

File: indexer/src/types/transaction.ts [hunk 13]
```diff
@@ -67,28 +72,133 @@ export function toEthTx({
...
+}
+
+/**
+ * Adds the isRunOutOfResources flag to the transaction result if the transaction
+ * was reverted due to running out of resources.
+ *
+ * @param receipt - The transaction receipt object.
+ * @param result - The transaction result object in Ethereum format.
+ */
+function flagResourceExhaustion(
```
Comment (tcoratger) [inline indexer/src/types/transaction.ts:123]:
```suggestion
function flagRunOutOfResources(
```

File: indexer/src/types/interfaces.ts [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+ */
+export type HexString = `0x${string}`
+
+/**
+ * Represents a request to convert a transaction to Ethereum transaction format.
+ */
+export interface ToEthTxRequest {
+  /** An Ethereum transaction. */
+  transaction: Transaction;
+  /** The Ethereum receipt corresponding to a reverted out of resources transaction. */
```
Comment (tcoratger) [inline indexer/src/types/interfaces.ts:23]:
```suggestion
  /** An Ethereum receipt. */
```

File: indexer/src/types/interfaces.ts [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+
+/**
+ * Represents a typed transaction to Ethereum transaction conversion request.
+ */
+export interface TypedTxToEthTx {
+  /** The typed transaction object. */
+  typedTransaction: TypedTransaction;
+  /** The Ethereum transaction object. */
+  transaction?: Transaction;
+  /** The Ethereum receipt corresponding to a reverted out of resources transaction. */
```
Comment (tcoratger) [inline indexer/src/types/interfaces.ts:51]:
```suggestion
  /** The Ethereum receipt. */
```

File: indexer/src/types/interfaces.ts [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+ * Represents the format for building a transaction in Ethereum format.
+ * This interface is used to construct an Ethereum formatted transaction
+ * from the provided typed transaction and its JSON representation.
+ */
+export interface BuildTransactionEthFormat {
+  /** The typed transaction object. */
+  typedTransaction: TypedTransaction;
+  /** The JSON representation of the transaction. */
+  jsonTx: JsonTx;
+  /** The Ethereum receipt corresponding to a reverted out of resources transaction. */
```
Comment (tcoratger) [inline indexer/src/types/interfaces.ts:71]:
```suggestion
  /** The Ethereum receipt. */
```

File: indexer/src/types/interfaces.ts [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+  blockNumber: PrefixedHexString;
+  /** The hash of the block in which the transaction was included, as a prefixed hex string. */
+  blockHash: PrefixedHexString;
+  /** Indicates if the transaction is in a pending block. */
+  isPendingBlock: boolean;
+  /** The chain id of the transaction. */
+  chainId: string | undefined;
+  /** The index of the transaction in the block. */
+  index: string;
+}
```
Comment (tcoratger) [inline indexer/src/types/interfaces.ts:83]:
```suggestion
}

```

File: indexer/src/types/transaction.ts [hunk 18]
```diff
@@ -67,28 +72,133 @@ export function toEthTx({
...
+ *
+ * @param typedTransaction - The typed transaction object.
+ * @param jsonTx - The JSON representation of the transaction.
+ * @returns - The chain ID as a string, or undefined if the chain ID cannot be determined.
+ *
+ * If the transaction is a legacy transaction that supports EIP-155 replay protection,
+ * the chain ID is calculated from the `v` value using the formula:
+ * v = 35 + 2 * chainId + yParity -> chainId = (v - 35) / 2
+ */
+function calculateChainId(typedTransaction: TypedTransaction, jsonTx: JsonTx): string | undefined {
```
Comment (tcoratger) [inline indexer/src/types/transaction.ts:85]:
```suggestion
function chainId(typedTransaction: TypedTransaction, jsonTx: JsonTx): string | undefined {
```

File: indexer/src/types/transaction.ts [hunk 19]
```diff
@@ -67,28 +72,133 @@ export function toEthTx({
...
+ * @param jsonTx - The JSON representation of the transaction.
  * @param blockNumber - The block number of the transaction in hex.
  * @param blockHash - The block hash of the transaction in hex.
  * @param isPendingBlock - Whether the block is pending.
- * @returns - The transaction in the Ethereum format, or null if the transaction is invalid.
+ * @param chainId - The chain ID for the transaction.
+ * @param index - The index of the transaction in the block.
+ * @returns - The transaction in the Ethereum format, or null if the transaction is not signed.
+ */
+function buildTransactionEthFormat({
```
Comment (tcoratger) [inline indexer/src/types/transaction.ts:146]:
```suggestion
function transactionEthFormat({
```

File: indexer/src/types/interfaces.ts [hunk 20]
```diff
@@ -0,0 +1,83 @@
...
+  /** The y parity of the signature. */
+  yParity?: string;
+  /** Indicates if the transaction is reverted due to running out of resources. */
+  isRunOutOfResources?: boolean;
+}
+
+/**
+ * Represents a typed transaction to Ethereum transaction conversion request.
+ */
+export interface TypedTxToEthTx {
```
Comment (tcoratger) [inline indexer/src/types/interfaces.ts:46]:
```suggestion
export interface TypedTransactionContext{
```

File: indexer/src/types/interfaces.ts [hunk 21]
```diff
@@ -0,0 +1,83 @@
...
+
+/**
+ * Represents a hexadecimal string with a `0x` prefix.
+ */
+export type HexString = `0x${string}`
+
+/**
+ * Represents a request to convert a transaction to Ethereum transaction format.
+ */
+export interface ToEthTxRequest {
```
Comment (tcoratger) [inline indexer/src/types/interfaces.ts:20]:
```suggestion
export interface TransactionContext {
```

File: indexer/src/types/interfaces.ts [hunk 22]
```diff
@@ -0,0 +1,83 @@
...
+  /** Indicates if the transaction is in a pending block. */
+  isPendingBlock: boolean;
+}
+
+/**
+ * Represents the format for building a transaction in Ethereum format.
+ * This interface is used to construct an Ethereum formatted transaction
+ * from the provided typed transaction and its JSON representation.
+ */
+export interface BuildTransactionEthFormat {
```
Comment (tcoratger) [inline indexer/src/types/interfaces.ts:66]:
```suggestion
export interface TransactionConversionInput {
```

**Comment by dznes** [comment]
i've made a mistake running the deno fmt into all files, other than the /index. So i made another PR with all the suggestions added.
Sorry about the inconvinience.
The new PR is in this link: https://github.com/kkrt-labs/kakarot-rpc/pull/1237

**Pull request closed without merging**